cmd/collect_signals/signal/signal.go

```diff
@@ -0,0 +1,241 @@
+package signal
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/iancoleman/strcase"
+)
+
+type Namespace string
+
+func (ns Namespace) String() string {
+	return string(ns)
+}
+
+const (
+	// nameSeparator is used to separate the namespace from the field name.
+	nameSeparator = '.'
+
+	// namespaceLegacy is an internal namespace used for fields that provide
+	// compatibility with the legacy python implementation.
+	namespaceLegacy Namespace = "legacy"
+
+	// The following constants are used while parsing the tags for the struct
+	// fields in a Set.
+	fieldTagName      = "signal"
+	fieldTagIgnore    = "-"
+	fieldTagLegacy    = "legacy"
+	fieldTagSeperator = ","
+)
+
+const (
+	NamespaceRepo   Namespace = "repo"
+	NamespaceIssues Namespace = "issues"
+)
+
+var (
+	// validName is used to validate namespace and field names.
+	validName = regexp.MustCompile("^[a-z0-9_]+$")
+
+	// valuerType caches the reflect.Type representation of the valuer
+	// interface.
+	valuerType = reflect.TypeOf((*valuer)(nil)).Elem()
+)
+
+type SupportedType interface {
+	~int | ~int8 | ~int16 | ~int32 | ~int64 |
+		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr |
+		~float32 | ~float64 |
+		~string | time.Time
+}
+
+// valuer is provides access to the field's value without needing to use
+// the concrete generic. It is also a marker interface used to identify the
+// Signal field in a struct.
+type valuer interface {
+	Value() any
+}
+
+type Field[T SupportedType] struct {
```

Comment:
More general question as someone very unfamiliar with Go generics: there's a lot of reflection used here to make this package work. 

What value does using generics give us here over just purely using reflection? Is it purely to make this easier to implement? 

The signal sets current require sets to be defined as:

```
type IssuesSet struct {
	UpdatedCount     Field[int]     `signal:"updated_issues_count,legacy"`
	ClosedCount      Field[int]     `signal:"closed_issues_count,legacy"`
	CommentFrequency Field[float64] `signal:"issue_comment_frequency,legacy"`
}
```

instead of just:

```
type IssuesSet struct {
	UpdatedCount    int     `signal:"updated_issues_count,legacy"`
	ClosedCount      int     `signal:"closed_issues_count,legacy"`
	CommentFrequency float64 `signal:"issue_comment_frequency,legacy"`
}
```

Which adds a layer of indirection -- what does this indirection buy us? If the goal is to just be able to extract key/value pairs from this, we could just do something similar to what encoding/json is doing right?

Reply:
The problem is that I want to support the idea of a field being "unset".

Using normal scalars like `int` is impossible for this. I could then use `*int`, and define my own `FieldInt` type, but then I might as well start using generics, e.g: `Field[T]` as I have.

The alternative is to allow `*int` and `int`, etc, and make the code handle it appropriately, but I suspect it is then easier to accidentally use an `int` and forget about the "unset" scenario. Trying to enforce pointer usage could only be done at runtime using reflection.

I feel that by using generics I can limit the use of reflection, and lean on the compiler to do a lot of the heavy lifting w.r.t type checking. I can also add functions to the `Field[T]` type as well that help with processing the data.

Reply:
Thanks, that makes sense. I didn't think about this supporting the "unset" case.